### PR TITLE
AJ-790 withErrorReporting and utils Typescript conversion

### PR DESCRIPTION
**Pull request by msilva-broad**
- The TS changes in AJ-790 is just another instance of many of some core shared curried functions and related utilities causing type issues.
- these changes do a partial upgrade of key files to improve the experience for the growing number of devs doing TS work.
- leveraging safeCurry and withHandlers helpers and related patterns.

This greatly improves type flow, letting local variables in consuming code infer the correct type signature for the methods/utils directly addressed in this PR

### CODE REVIEW

File: src/libs/utils.ts [hunk 1]
```diff
@@ -348,27 +360,34 @@ export const commaJoin = (list, conjunction = 'or') => {
...
+  const lookup: [string, number][] = [
     ['P', 2 ** 50],
     ['T', 2 ** 40],
     ['G', 2 ** 30],
     ['M', 2 ** 20],
     ['K', 2 ** 10]
-  ].find(([_p, d]) => bytes >= d)
-  return `${(bytes / divisor).toPrecision(3)} ${prefix}iB`
+  ]
+  const maybeLookup = lookup.find(([_p, d]: [string, number]) => bytes >= d)
```
Comment (msilva-broad) [inline src/libs/utils.ts:380]: this is to fix an unhandled edge-case which was exposed by Typescript complaining about this code's previous form.  But let me add some unit tests for this...

File: src/pages/workspaces/workspace/analysis/file-utils.ts [hunk 2]
```diff
@@ -65,17 +66,16 @@ export const useAnalysisFiles = (): AnalysisFileStore => {
   const [analyses, setAnalyses] = useState<AnalysisFile[]>([])
   const [pendingCreate, setPendingCreate] = useLoadedData<true>()
 
-  const refresh: () => Promise<void> = _.flow(
-    // @ts-expect-error
+  const refresh = withHandlers([
```
Comment (msilva-broad) [inline src/pages/workspaces/workspace/analysis/file-utils.ts:69]: the cast is no longer needed, and the type is reliably inferred now.... AWESOME!

File: src/libs/error.ts [hunk 3]
```diff
@@ -13,18 +15,29 @@ export const reportError = async (title, obj) => {
   notify('error', title, { detail: await (obj instanceof Response ? obj.text() : obj) })
 }
 
+export type ErrorCallback = (error: unknown) => void | Promise<void>
+const withErrorHandlingFn = <R, F extends AnyPromiseFn>(
```
Comment (msilva-broad) [inline src/libs/error.ts:19]: breaking the function out from the currying step below is key to being able to properly type the function and leverage earlier work to let these types flow well through safeCurry and optionally withHandlers utility on the consuming side.

File: src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts [hunk 4]
```diff
@@ -19,6 +18,7 @@ import {
 import { AppTool, getToolLabelFromFileExtension, ToolLabel, tools } from 'src/pages/workspaces/workspace/analysis/tool-utils'
 import { asMockedFn } from 'src/testing/test-utils'
 
+import { reportError } from '../../../../../libs/error'
```
Comment (msilva-broad) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisModal.test.ts:21]: curious... yarn lint cmd auto-fixed these imports for correct order, and did this relative pathing change here and in a few other files...

File: src/libs/utils.ts (hunk 1 above)
Comment (msilva-broad) [reply src/libs/utils.ts:380]: unit tests added

File: src/libs/utils.test.ts [hunk 5]
```diff
@@ -136,3 +136,27 @@ describe('isValidWsExportTarget', () => {
...
   })
 })
+
+describe('formatBytes', () => {
+  it('handles GB', () => {
+    // Act
+    const result = formatBytes(40000000000)
+
+    // Assert
+    expect(result).toBe('37.3 GiB')
```
Comment (msilva-broad) [inline src/libs/utils.test.ts:146]: what's up with the little 'i' in GiB?  This looks pretty awkward for the other results like MiB below...

File: src/libs/error.ts [hunk 6]
```diff
@@ -13,18 +15,29 @@ export const reportError = async (title, obj) => {
...
+
 /**
  * Invoke the `callback` with any error thrown when evaluating the async `fn` with `...args`.
  */
-export const withErrorHandling = _.curry((callback, fn) => async (...args) => {
-  try { return await fn(...args) } catch (error) { await callback(error) }
-})
+export const withErrorHandling = safeCurry(withErrorHandlingFn)
 
+// TODO type this properly
```
Comment (s-rubenstein) [inline src/libs/error.ts]: Is there a ticket for the TODO? Should it be linked here?

File: src/libs/utils.ts [hunk 7]
```diff
@@ -3,24 +3,27 @@ import { differenceInCalendarMonths, differenceInSeconds, parseJSON } from 'date
...
 import * as qs from 'qs'
 import { div, span } from 'react-hyperscript-helpers'
+import { AnyPromiseFn, GenericPromiseFn } from 'src/libs/type-utils/general-types'
+import { safeCurry } from 'src/libs/type-utils/lodash-fp-helpers'
 import { v4 as uuid } from 'uuid'
 
 import { getCloudProviderFromWorkspace } from './workspace-utils'
 
 
+// TODO: add good typing (remove any's)
```
Comment (s-rubenstein) [inline src/libs/utils.ts]: Similarly, is there a ticket for this TODO? Can we link it here if so?

**Review by s-rubenstein — APPROVED** [review]
Overall LGTM, I would just ticket some of the TODOs?

File: src/libs/error.ts (hunk 6 above)
Comment (msilva-broad) [reply src/libs/error.ts]: I can create one on UIE board and link here and in similar comments.

Comment (msilva-broad) [reply src/libs/error.ts]: I'll add this ticket link to the TODOs before merge:

https://broadworkbench.atlassian.net/browse/UIE-67

File: src/libs/utils.test.ts (hunk 5 above)
Comment (jdcanas) [reply src/libs/utils.test.ts:146]: Its actually a different unit. Base 10 instead of base 2 measurement of bytes. Google's choice, not ours (in terms of how they size their cloud resources).

**Review by jdcanas — APPROVED** [review]

File: src/libs/ajax/GoogleStorage.ts [hunk 8]
```diff
@@ -79,7 +79,6 @@ const getServiceAccountToken: (googleProject: string, token: string) => Promise<
   return res.json()
 }, {
   expires: 1000 * 60 * 30,
-  // @ts-expect-error
```
Comment (aaronkanzer) [inline src/libs/ajax/GoogleStorage.ts:82]: Is this intentional?
